frontend/src/metabase/core/components/Slider/Slider.tsx

```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
-      ((Math.min(...value) - rangeMin) / totalRange) * 100,
+      ((Math.min(...value) - min) / totalRange) * 100,
       (Math.abs(value[1] - value[0]) / totalRange) * 100,
     ];
-  }, [value, rangeMin, rangeMax]);
+  }, [value, min, max]);
 
-  const handleChange = useCallback(
+  const handleInput = useCallback(
     (event: ChangeEvent<HTMLInputElement>, valueIndex: number) => {
       const changedValue = [...value];
       changedValue[valueIndex] = Number(event.target.value);
-      onChange(changedValue);
+      setValue(changedValue);
     },
-    [value, onChange],
+    [value, setValue],
   );
 
-  const sortValues = useCallback(() => {
-    if (value[0] > value[1]) {
-      const sortedValues = [...value].sort();
-      onChange(sortedValues);
-    }
+  const handleChange = useCallback(() => {
+    const sortedValues = value[1] < value[0] ? [...value].sort() : value;
+    onChange(sortedValues);
   }, [value, onChange]);
 
+  const [minValue, maxValue] = useMemo(
+    () =>
+      value.every(n => !isNaN(n))
+        ? [Math.min(...value), Math.max(...value)]
+        : value,
+    [value],
+  );
+
   return (
-    <SliderContainer>
+    <SliderContainer
+      onMouseEnter={() => setIsHovering(true)}
+      onMouseLeave={() => setIsHovering(false)}
+    >
       <SliderTrack />
-      <ActiveTrack left={beforeRange} width={rangeWidth} />
+      <ActiveTrack
+        style={{ left: `${beforeRange}%`, width: `${rangeWidth}%` }}
+      />
+      <SliderTooltip
+        data-testid="min-slider-tooltip"
+        style={{
+          left: getTooltipPosition(beforeRange),
+          opacity: isHovering ? 1 : 0,
+        }}
+      >
+        {hasDecimal(step) ? minValue.toFixed(2) : minValue}
+      </SliderTooltip>
       <SliderInput
         type="range"
         aria-label="min"
         value={value[0]}
-        onChange={e => handleChange(e, 0)}
-        onMouseUp={sortValues}
-        min={rangeMin}
-        max={rangeMax}
+        onChange={e => handleInput(e, 0)}
+        onMouseUp={handleChange}
+        onKeyUp={handleChange}
+        min={min}
+        max={max}
         step={step}
       />
+      <SliderTooltip
+        data-testid="max-slider-tooltip"
+        style={{
+          left: getTooltipPosition(beforeRange + rangeWidth),
+          opacity: isHovering ? 1 : 0,
+        }}
+      >
+        {hasDecimal(step) ? maxValue.toFixed(2) : maxValue}
+      </SliderTooltip>
       <SliderInput
         type="range"
         aria-label="max"
         value={value[1]}
-        onChange={e => handleChange(e, 1)}
-        onMouseUp={sortValues}
-        min={rangeMin}
-        max={rangeMax}
+        onChange={e => handleInput(e, 1)}
+        onMouseUp={handleChange}
+        onKeyUp={handleChange}
+        min={min}
+        max={max}
         step={step}
       />
     </SliderContainer>
   );
 };
 
+const getTooltipPosition = (basePosition: number) =>
+  `calc(${basePosition}% + ${11 - basePosition * 0.18}px)`;
```

Comment:
This is a gnarly bit of math to line the tooltip up with the center of the drag handle

Reply:
#### I don't think it's worth changing considering how your simple calculation is almost correct like 99.9%

Nit: I believe this math is somewhat off. Like it's not totally off, but just by a bit. Basically, the width of the tooltip container is larger than the length the slider thumb can travel by the thumb width.

That means we need to scale the track 100% down to the thumb travel length. And that's not possible via CSS. From my research, we can't put any percentage as a divisor.

So what I came up with is to put tooltips in a new container and make the container the size of the slider thumb travel length. Then we can easily match the position of the tooltip with the slider thumb.
https://www.loom.com/share/33ce4de78604438d81343b2dde0cc04f

Reply:
your way seems better, and with styled components, I don't think it actually makes the code harder to read.

You also got me looking at things WAY zoomed in so I made a couple other geometry tweaks. html range inputs are weird.